persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DsePersistence.java

```diff
@@ -269,26 +269,40 @@ public Authenticator getAuthenticator() {
       boolean isTracingRequested,
       long queryStartNanoTime) {
     try {
-      org.apache.cassandra.service.QueryState internalState = Conversion.toInternal(state);
-      org.apache.cassandra.cql3.QueryOptions internalOptions = Conversion.toInternal(options);
-
       return Conversion.toFuture(
           Single.defer(
               () -> {
                 try {
-                  final UUID tracingId =
-                      beginTraceQuery(
-                          cql, internalState, internalOptions, customPayload, isTracingRequested);
+                  org.apache.cassandra.service.QueryState internalState =
+                      Conversion.toInternal(state);
+                  org.apache.cassandra.cql3.QueryOptions internalOptions =
+                      Conversion.toInternal(options);
 
                   checkIsLoggedIn(internalState);
 
                   CQLStatement statement = QueryProcessor.parseStatement(cql, internalState);
+
+                  // Please note that this needs to happen _before_ the beginTraceQuery, because
+                  // when we add bound values to the trace, we rely on the values having been
+                  // re-ordered by the following prepare (if named values were used that is).
                   internalOptions.prepare(statement.getBindVariables());
 
+                  final UUID tracingId =
+                      beginTraceQuery(
+                          cql, internalState, internalOptions, customPayload, isTracingRequested);
                   return processStatement(
-                      statement, state, options, customPayload, queryStartNanoTime, tracingId);
+                      statement,
+                      state,
+                      options,
+                      internalState,
+                      internalOptions,
+                      customPayload,
+                      queryStartNanoTime,
+                      tracingId);
                 } catch (Exception e) {
-                  return Single.error(Conversion.handleException(e));
+                  return Tracing.instance
```

Comment:
I know that this is happening inside of the async flow, but I wonder if we shouldn't make a `stop()` (blocking) instead of the async version?
If there are an exception and the client will not chain the operations properly on the returned `CompletableFuture` we may risk the situation in which the session will not be closed when needed, wdyt?